### DISCUSSION TITLE
Cannot delete or update a parent row: a foreign key constraint fails

### DISCUSSION THREAD

**Discussion by alan345** [prisma/prisma#4061] [Q&A] [answered]
Using this doc: https://www.prisma.io/docs/guides/database-workflows/foreign-keys/mysql
I want to to add a simple join between 2 tables. `PromoCode` and `Invoice` tables already exists. 

```prisma
model PromoCode {
  id         String    @id @default(cuid())
  code       String
  invoices   Invoice[]
}

model Invoice {
  id               String                  @id @default(cuid())
  promoCodeId      String?
  promoCode        PromoCode?  @relation(fields: [promoCodeId], references: [id])
}
```

I am running this SQL snippet to join the 2 tables:
```sql
ALTER TABLE `Invoice` ADD COLUMN `promoCodeId` char(30) CHARACTER SET utf8;

ALTER TABLE Invoice 
ADD CONSTRAINT `Invoice_ibfk_8`
FOREIGN KEY (`promoCodeId`) REFERENCES `PromoCode`(`id`);
```

It works Well, and I can add data.

Now If I delete a promoCode, It drops me this error:
```
[Error code:1451 SQL state:23000] Cannot delete or update a parent row: a foreign key constraint fails (`nacho@prod`.`invoice`, CONSTRAINT `Invoice_ibfk_8` FOREIGN KEY (`promoCodeId`) REFERENCES `promocode` (`id`))
```

I cannot figure out what is wrong as 1 `invoice` can have 0 or 1 `PromoCode`, So `PromoCode` should be possible to be deleted.

What should be the correct SQL snippet to join the 2 tables?

**Comment by ryands17** [comment, marked as answer]
Are you using introspection or Prisma migrate to create the tables?

**Reply by alan345** [reply]
I am using pure SQL:
First, I run this snippet:
```sql
ALTER TABLE `Invoice` ADD COLUMN `promoCodeId` char(30) CHARACTER SET utf8;

ALTER TABLE Invoice 
ADD CONSTRAINT `Invoice_ibfk_8`
FOREIGN KEY (`promoCodeId`) REFERENCES `PromoCode`(`id`);
```
and manually change `schema.prisma`

```prisma
model PromoCode {
  id         String    @id @default(cuid())
  code       String
  invoices   Invoice[]
}

model Invoice {
  id               String                  @id @default(cuid())
  promoCodeId      String?
  promoCode        PromoCode?  @relation(fields: [promoCodeId], references: [id])
}
```

I cannot run 
- `migrate` because of https://github.com/prisma/prisma/discussions/4079
- `introspection` because of https://github.com/prisma/prisma/discussions/3976

**Reply by ryands17** [reply]
Could you try this SQL:

```sql
CREATE TABLE "PromoCode" (
    "id" TEXT NOT NULL,
    "code" TEXT NOT NULL,
    PRIMARY KEY ("id")
)

CREATE TABLE "Invoice" (
    "id" TEXT NOT NULL,
    "promoCodeId" TEXT,
    FOREIGN KEY ("promoCodeId") REFERENCES "PromoCode"("id") ON DELETE SET NULL ON UPDATE CASCADE,
PRIMARY KEY ("id")
```

By creating the above tables and using the schema that you have provided, promoCode deletion works fine.
I'm on version _2.9.0_ of `@prisma/client` and `@prisma/cli`.

**Reply by alan345** [reply]
Yes it works fine for me too. 
The issue is when the tables `PromoCode` and `Invoice` already exit for a long time.
And now, I have to add a new join between those tables. I cannot use `CREATE TABLE`, so I used `ALTER TABLE`:

```sql
ALTER TABLE `Invoice` ADD COLUMN `promoCodeId` char(30) CHARACTER SET utf8;

ALTER TABLE Invoice 
ADD CONSTRAINT `Invoice_ibfk_8`
FOREIGN KEY (`promoCodeId`) REFERENCES `PromoCode`(`id`);
```
 
I guess I am doing something wrong here.

**Reply by ryands17** [reply]
Maybe add the CASCADE rules as well in your alter table commands?

**Reply by alan345** [reply]
I I am using  `ON DELETE CASCADE`, If I delete a PromoCode, the Invoice will also be deleted. 
The logic is: 1 invoice can have 0 or 1 PromoCode.

So if I delete the promoCode, the Invoice should now have 0 PromoCode. (instead of 1)

**Reply by ryands17** [reply]
> The logic is: 1 invoice can have 0 or 1 PromoCode.

In this case, it wouldn't be necessary to delete the Invoice once you delete the Promocode. The invoice will remain and you will have 0 promocode (which will be `NULL`). I guess that's the reason it's failing.

**Reply by alan345** [reply]
I Agree, It is what I expect! And when I do 
```
      promoCode = await ctx.prisma.promoCode.delete({
        where: {
          id: args.where.id,
        },
      })
```

I have this error: 
```
   49 async deletePromoCode(parent: any, args: any, ctx: Ctx, info: any) {
   50   let promoCode
   51   try {
→  52     promoCode = await ctx.prisma.promoCode.delete(
  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: ForeignKeyConstraintViolation { constraint: Fields(["promoCodeId"]) } })
    at $w.request (/Users/alan/Documents/app/nacho/server/node_modules/@prisma/client/runtime/src/runtime/getPrismaClient.ts:1270:15)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  clientVersion: '2.9.0'
```

**Reply by ryands17** [reply]
Yes but for that you would need to set `ON DELETE SET NULL` only then you will be able to satisfy the above constraint.

**Reply by alan345** [reply]
`ON DELETE SET NULL ` works! Thanks. And its in the [doc](https://www.prisma.io/docs/guides/database-workflows/cascading-deletes/postgresql#5-create-two-tables-with-a-foreign-key-and-set-null-deletion-behavior)!

I guess this condition will be overwritten in my databse when I will run `prisma migrate`?
```
ALTER TABLE `Invoice` ADD COLUMN `promoCodeId` char(30) CHARACTER SET utf8;

ALTER TABLE Invoice 
ADD CONSTRAINT `Invoice_ibfk_8`
FOREIGN KEY (`promoCodeId`) REFERENCES `PromoCode`(`id`)
ON DELETE SET NULL;
```
Thanks

**Reply by ryands17** [reply]
No mostly it shouldn't be overwritten. Could you try and check just to be sure?

**Reply by alan345** [reply]
I will do it and let you know as soon as I will be able to do it! :) https://github.com/prisma/prisma/discussions/4079

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
